### ISSUE TITLE
This is a awesome idea, some feedback

### ISSUE THREAD

**Issue by cancerberoSgx** [LeDDGroup/typescript-transform-macros#2]
I recently learn haxe and macros is a powerful tool, Let me give you a couple of thoughts:

 * I would expect macros to be a preprocessor TypeScript-> TypeScript transformation. (currently they seem to be implemented as transformer TS->JS
 * I know tsc currently doesn't allow to register a transformer as plugin . (so it's better to just build a preprocessor TS->TS tool - doesn't need to be a plugin)
 * Also, as in haxe, TypeScript AST own API types can be used by the user to reference / select / build code. There are nice libraries for select nodes / manipulate the AST and concretely declare code programmatically (check the tool at the bottom/left of this tool https://ts-ast-viewer.com/). On the other side, this could limit things you can do, for example, adding a new method - how the type checking would work in that case... (just thinking aloud- )
 * I personally would not install ttypescript or anything different than the official thing for generating js... Is it really worthwhile even in the case that the output is js ? isn't make more sense to have an intermediary  TS->TS tool instead ? 

In summary I think this is an awesome idea, feel free to close this / I would be glad to keep discussing on this. thanks!

**Comment by danielpza** [comment]
Hi, yes, ideas and feedback are appreciated.

First, this project is a TS->TS transformer, but the processes are mixed up as one.
You could use the typescript api to apply this transformer only and received typescript code from typescript code, eg see https://github.com/LeDDGroup/typescript-transform-macros/blob/master/src/index.test.ts#L20, in the tests the code is transformed and not transpiled to js.

As of ttypescript it seems to be one of the few alternatives (some plugins for typescript in webpack also allow this) that allows custom transformers. I've used it for a while and it is just a wrapper for typescript plus the transform option, it has all the features typescript has. I would use typescript directly but they don't seem/plan to support it, there are some open issues on github/typescript on the subject, https://github.com/microsoft/TypeScript/issues/14419

I've not found any tool to apply typescript transformers only (ie not transpiling to js) at the moment.

**Comment by cancerberoSgx** [comment]
> You could use the typescript api to apply this transformer only and received typescript code from typescript code, eg see https://github.com/LeDDGroup/typescript-transform-macros/blob/master/src/index.test.ts#L20, in the tests the code is transformed and not transpiled to js.

Oh, from the readme that's not clear , my bad, I got exited about the project and talk without even trying it.

I see you already tackled the common use case for adding new methods to existing classes. Another cool use case could be being able to inline a function call. An also embedding a string, array , files, etc. I think having more examples like these could boost this project "popularity" since for javascripters the concept of macros could be unknown.

Question: Can't the MACRO function declaration be defined in the library itself as a global? I think it can and that will save users o declare it, just import it (unless there is really need for that)

Also, an alternative setup could be just providing a CLI tool that translate an entire project. Perhaps that is a more straight forward experience than installing ttypescript and editing .json. 
Libraries like  ts-morph have easy to use APIs to load, manipulate and transform a ts project from its tsconfig.json file . There are other cool ones for translate code as string to TS AST node factory function call expressions, and others to query/select nodes from the AST. Let me take a look first and the I will suggest some (probably you already consider them, but just in case).

Thanks! I have more comments but I will test before talking ;) and come back to you

**Comment by danielpza** [comment]
> Oh, from the readme that's not clear , my bad, I got exited about the project and talk without even trying it.

no worries, the readme and the whole ecosystem is not clear enough.

> Question: Can't the MACRO function declaration be defined in the library itself as a global? I think it can and that will save users o declare it, just import it (unless there is really need for that)

Yes, it could, and there are some other features that would be cool to have, but currently I lack motivation to continue developing the project because I don't use it, and I've got little feedback from the community. I started this as a fun project, and most of the code is messy and due for a refactoring, but pull requests (and ideas and feedback such as yours) are welcome.

**Comment by danielpza** [comment]
Also, here are the test input and output (both in typescript):

https://github.com/LeDDGroup/typescript-transform-macros/blob/master/src/__fixtures/input.ts

https://github.com/LeDDGroup/typescript-transform-macros/blob/master/src/__snapshots__/index.test.ts.snap#L4-L39
